### PR TITLE
Taking new ATParser and FileHandle in use

### PR DESCRIPTION
**Pull request by hasnainvirk**
The driver is modified to take advantage of new in-tree ATParser and
FileHandle.

### CODE REVIEW

**Comment by hasnainvirk** [comment]
@geky @sarahmarshy @SeppoTakalo This will go in after https://github.com/ARMmbed/mbed-os/pull/4119 is merged.

@sarahmarshy You might need to fast-track this https://github.com/ARMmbed/esp8266-driver/pull/23.
I can then rebase and test again with my changes.

File: ESP8266/ESP8266.cpp [hunk 1]
```diff
@@ -16,14 +16,22 @@
...
+
 ESP8266::ESP8266(PinName tx, PinName rx, bool debug)
-    : _serial(tx, rx, 1024), _parser(_serial)
-    , _packets(0), _packets_end(&_packets)
+    :  _packets(0), _packets_end(&_packets)
 {
-    _serial.baud(115200);
-    _parser.debugOn(debug);
+    _fh = new BufferedSerial(tx, rx, BAUDRATE);
+    _parser =  new ATParser(*_fh, ATPARSER_BUFSIZE, ATPARSER_TIMEOUT, debug);
```
Comment (geky) [inline ESP8266/ESP8266.cpp]: Why are these dynamically allocated now? They can just be declared in the class, virtual inheritance will still work if passed by address.

File: ESP8266Interface.h [hunk 2]
```diff
@@ -161,6 +161,12 @@ class ESP8266Interface : public NetworkStack, public WiFiInterface
      */
     using NetworkInterface::add_dns_server;
 
+    /** Check if connected to the network
+     *
+     *  @return         true if connected, false otherwise
+     */
+    bool isConnected(void);
```
Comment (geky) [inline ESP8266Interface.h]: I wouldn't add this function unless it is also added to the network interface, otherwise we may run into name conflicts when we do

File: ESP8266/ESP8266.cpp (hunk 1 above)
Comment (geky) [inline ESP8266/ESP8266.cpp]: Also why are you dereferencing `_fh`? Won't that copy the entire BufferedSerial class?

File: ESP8266/ESP8266.cpp [hunk 3]
```diff
@@ -279,28 +287,24 @@ bool ESP8266::close(int id)
 
 void ESP8266::setTimeout(uint32_t timeout_ms)
 {
-    _parser.setTimeout(timeout_ms);
+    _parser->setTimeout(timeout_ms);
 }
 
-bool ESP8266::readable()
-{
-    return _serial.readable();
```
Comment (geky) [inline ESP8266/ESP8266.cpp]: Couldn't you replace these with the FileHandle::poll function?

File: ESP8266/ESP8266.cpp [hunk 4]
```diff
@@ -279,28 +287,24 @@ bool ESP8266::close(int id)
...
+    if ((events & MBED_POLLIN )|| (events & MBED_POLLOUT)) {
+        callback(_callback);
+    }
 }
-
 void ESP8266::attach(Callback<void()> func)
 {
-    _serial.attach(func);
+    _callback = func;
+    _fh->attach(callback(this, &ESP8266::stream_event_cb));
```
Comment (geky) [inline ESP8266/ESP8266.cpp]: Do we care about the type of events? The sigio function does not. Can we skip the `ESP8266::stream_event_cb` function?

**Comment by geky** [comment]
#23 should be coming in soon, just a few style changes left. There seems to be quite a few unnecessary changes (now using dynamic memory)?

Sidenote for self: This will need to wait for https://github.com/ARMmbed/mbed-os/pull/4119 to be merged

File: ESP8266/ESP8266.cpp (hunk 1 above)
Comment (hasnainvirk) [inline ESP8266/ESP8266.cpp]: Dynamic allocation means, we can clean up memory later. _fh is a pointer  and hence have a value which is a memory address .. *_fh means take the value at _fh ,i.e., memory address .. So a pointer is being passed here

File: ESP8266Interface.h (hunk 2 above)
Comment (hasnainvirk) [inline ESP8266Interface.h]: This can be removed, it was just a convenient function .. There is nothing wrong with having something as fundamental as checking for a connection .. However, I am not so keen on it .. Can be removed

File: ESP8266/ESP8266.cpp (hunk 3 above)
Comment (hasnainvirk) [inline ESP8266/ESP8266.cpp]: poll() have a dummy implementation in FileHandle , the derived class should implement that. mbed_poll() on the other hand which is exposed by the FileHandle header file is blocking in nature.
I did not find any piece of code using these APIs that's why I removed it. You can implement poll() if necessary .. Or if you have something in your mind, go ahead and share with me

File: ESP8266/ESP8266.cpp (hunk 4 above)
Comment (hasnainvirk) [inline ESP8266/ESP8266.cpp]: _fh->attach requires a callback function with 'short events' as an argument .. That's why I had to wrap it

File: ESP8266/ESP8266.cpp (hunk 3 above)
Comment (kjbracey-arm) [inline ESP8266/ESP8266.cpp]: What should the readable function even mean for a network interface? Doesn't seem to make sense to me as a public API.

Comment (geky) [inline ESP8266/ESP8266.cpp]: Hmm, that's fair, ok we can drop it 👍

File: ESP8266/ESP8266.cpp (hunk 4 above)
Comment (geky) [inline ESP8266/ESP8266.cpp]: Ah gotcha, that makes sense. Should other events such as error trigger the sigio though?

File: ESP8266/ESP8266.cpp (hunk 1 above)
Comment (geky) [inline ESP8266/ESP8266.cpp]: I realized ATParser is taking a reference, I moved that issue over to [here](https://github.com/ARMmbed/mbed-os/pull/4119/files#r113273995).

As for the dynamic memory allocation, I'm still confused on what that is accomplishing. Do you need to clean up the memory outside of the life of the enclosing class? As it is the dynamic memory will add a layer of indirection and just make it harder to statical determine your module's memory consumption.

File: ESP8266/ESP8266.cpp (hunk 4 above)
Comment (hasnainvirk) [inline ESP8266/ESP8266.cpp]: No there is a check in the routine, only if event is POLLIN or POLLOUT, it calls the given callback

File: ESP8266/ESP8266.cpp (hunk 1 above)
Comment (hasnainvirk) [inline ESP8266/ESP8266.cpp]: There will be cases where this WifiInterface will be used in conjunction with other XYZInterfaces .. In those scenarios, it will be preferable to shutdown something and completely get rid of all memory allocations relating to that entity .. I'm thinking about multi radio systems , wifi-cellular-BT all togather. An interface does its job and then gets shutdown and makes space for other processes to run. For a standalone wifi example your solution is great, but we should also be thinking about real life applications and use cases where indeed lifetime of this interface might cut short.

File: ESP8266/ESP8266.cpp (hunk 4 above)
Comment (geky) [inline ESP8266/ESP8266.cpp]: But should it be masking out errors? Before the sigio function was triggered for any events the owner would need to be woken up for, including errors. Eventually I would expect the sockets to have the same poll function, but for now I'm wondering if this should just be a pass-through.

File: ESP8266/ESP8266.cpp (hunk 1 above)
Comment (geky) [inline ESP8266/ESP8266.cpp]: In that situation, wouldn't you want to be dynamically allocating the ESP8266Interface itself?

**Comment by hasnainvirk** [comment]
@geky @sarahmarshy Please review again

File: ESP8266/ESP8266.cpp [hunk 5]
```diff
@@ -215,7 +220,7 @@ void ESP8266::_packet_handler()
     uint32_t amount;
 
     // parse out the packet
-    if (!_parser.recv(",%d,%d:", &id, &amount)) {
+    if (!_parser.recv(",%d,%" SCNu32 ":" , &id, &amount)) {
```
Comment (sarahmarshy) [inline ESP8266/ESP8266.cpp]: Can we just use `%d` here? It is unnecessary, and the type macros make it harder to read. If there are warnings, maybe use `%ld`?

File: ESP8266/ESP8266.cpp [hunk 6]
```diff
@@ -294,27 +299,29 @@ void ESP8266::setTimeout(uint32_t timeout_ms)
...
+    _serial.sigio(func);
 }
 
 bool ESP8266::recv_ap(nsapi_wifi_ap_t *ap)
 {
-    int sec;
-    bool ret = _parser.recv("+CWLAP:(%d,\"%32[^\"]\",%hhd,\"%hhx:%hhx:%hhx:%hhx:%hhx:%hhx\",%d", &sec, ap->ssid,
-                            &ap->rssi, &ap->bssid[0], &ap->bssid[1], &ap->bssid[2], &ap->bssid[3], &ap->bssid[4],
-                            &ap->bssid[5], &ap->channel);
+    // Newlib-nano doesn't support %hhx or SCNu8 scan formats,
```
Comment (geky) [inline ESP8266/ESP8266.cpp]: This seems like an unrelated change, should be in a different pr. Do we actually support newlib-nano? there are some other issues with the newlib-nano stdlib such as incorrect scanf return codes that could make this library unstable.

File: ESP8266/ESP8266.cpp (hunk 5 above)
Comment (hasnainvirk) [inline ESP8266/ESP8266.cpp]: Not really. ld means long signed int. Proper way of doing it is this. I don't see any other proper alternatives.

File: ESP8266/ESP8266.cpp (hunk 6 above)
Comment (hasnainvirk) [inline ESP8266/ESP8266.cpp]: Yes we do use new-lib nano. This is a minor change, can be tagged along here. I don't think this needs a separate PR

File: ESP8266/ESP8266.cpp (hunk 5 above)
Comment (geky) [inline ESP8266/ESP8266.cpp]: If we want to be pedantic, we should change `ESP8266::send` to take in `size_t`. I'm not sure SCNu32 adds anything useful here.

Comment (hasnainvirk) [inline ESP8266/ESP8266.cpp]: That would be an API change.

Comment (hasnainvirk) [inline ESP8266/ESP8266.cpp]: Well, probably we can do an API change here as it is internal to us

Comment (hasnainvirk) [inline ESP8266/ESP8266.cpp]: But that still wont help us regarding hhx format specifiers.

Comment (geky) [inline ESP8266/ESP8266.cpp]: Well, either way, the function's argument does not have significance as a uint32_t. I would keep the old string to be consistent with the rest of the driver.

Comment (hasnainvirk) [inline ESP8266/ESP8266.cpp]: Old string throws out ugly warnings. I would still argue for a cleaner code and output.

File: ESP8266/ESP8266.cpp (hunk 6 above)
Comment (geky) [inline ESP8266/ESP8266.cpp]: %hhx missing from newlib nano is concerning and I'm not really sold on supporting newlib nano for this driver since we rely on the stdlib's formating so much. But we can move forward with this if you want.

File: ESP8266/ESP8266.cpp (hunk 5 above)
Comment (geky) [inline ESP8266/ESP8266.cpp]: Yeah, the %hhx formats missing from newlib-nano is concerning

File: ESP8266/ESP8266.cpp (hunk 6 above)
Comment (kjbracey-arm) [inline ESP8266/ESP8266.cpp]: As far as I'm aware, we're still trying to use newlib-nano in GCC builds, which is pretty annoying. Can't print 64-bit numbers either.

But if so, I'm not sure how this function ever worked in a GCC build. That does puzzle me.

File: ESP8266/ESP8266.cpp (hunk 5 above)
Comment (kjbracey-arm) [inline ESP8266/ESP8266.cpp]: If it's size_t, then it wants to be "%zu", which newlib-nano also doesn't support. Only pedantic thing that would be supported is taking unsigned long and using "%lu".

**Comment by SeppoTakalo** [comment]
ATCmdparser is already taken into use so closing this one.
Not sure about how relevant the filehandle changes were.
